src/main/java/jcifs/smb/NtlmContext.java

```diff
@@ -0,0 +1,505 @@
+     */
+    @Override
+    public boolean isPreferredMech ( ASN1ObjectIdentifier mechanism ) {
+        return this.auth.isPreferredMech(mechanism);
+    }
+
+
+    @Override
+    public boolean isEstablished () {
+        return this.isEstablished;
+    }
+
+
+    /**
+     * @return the server's challenge
+     */
+    public byte[] getServerChallenge () {
+        return this.serverChallenge;
+    }
+
+
+    @Override
+    public byte[] getSigningKey () {
+        return this.masterKey;
+    }
+
+
+    @Override
+    public String getNetbiosName () {
+        return this.netbiosName;
+    }
+
+
+    /**
+     * @param targetName
+     *            the target's SPN
+     */
+    public void setTargetName ( String targetName ) {
+        this.targetName = targetName;
+    }
+
+
+    @Override
+    public byte[] initSecContext ( byte[] token, int offset, int len ) throws SmbException {
+        switch ( this.state ) {
+        case 1:
+            return makeNegotiate(token);
+        case 2:
+            return makeAuthenticate(token);
+        default:
+            throw new SmbException("Invalid state");
+        }
+    }
+
+
+    protected byte[] makeAuthenticate ( byte[] token ) throws SmbException {
+        try {
+            Type2Message msg2 = new Type2Message(token);
+
+            if ( log.isTraceEnabled() ) {
+                log.trace(msg2.toString());
+                log.trace(Hexdump.toHexString(token));
+            }
+
+            this.serverChallenge = msg2.getChallenge();
+
+            if ( this.requireKeyExchange ) {
+                if ( this.transportContext.getConfig().isEnforceSpnegoIntegrity() && ( !msg2.getFlag(NtlmFlags.NTLMSSP_NEGOTIATE_KEY_EXCH)
+                        || !msg2.getFlag(NtlmFlags.NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY) ) ) {
+                    throw new SmbUnsupportedOperationException("Server does not support extended NTLMv2 key exchange");
+                }
+
+                if ( !msg2.getFlag(NtlmFlags.NTLMSSP_NEGOTIATE_128) ) {
+                    throw new SmbUnsupportedOperationException("Server does not support 128-bit keys");
+                }
+            }
+
+            this.ntlmsspFlags &= msg2.getFlags();
+            Type3Message msg3 = createType3Message(msg2);
+            msg3.setupMIC(this.type1Bytes, token);
+
+            byte[] out = msg3.toByteArray();
+
+            if ( log.isTraceEnabled() ) {
+                log.trace(msg3.toString());
+                log.trace(Hexdump.toHexString(token));
+            }
+
+            this.masterKey = msg3.getMasterKey();
+
+            if ( this.masterKey != null && ( this.ntlmsspFlags & NtlmFlags.NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY ) != 0 ) {
+                initSessionSecurity(msg3.getMasterKey());
+            }
+
+            this.isEstablished = true;
+            this.state++;
+            return out;
+        }
+        catch ( SmbException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)